core/src/main/scala/org/bykn/bosatsu/Normalization.scala

```diff
@@ -0,0 +1,397 @@
+   * or there are no lambda variables used in the linked expressions
+   */
+  def maxLambdaVar: Option[Int]
+}
+
+object NormalExpression {
+  case class App(fn: NormalExpression, arg: NormalExpression)
+  extends NormalExpression {
+    val maxLambdaVar = (fn.maxLambdaVar.toList ++ arg.maxLambdaVar.toList)
+      .reduceLeftOption(Math.max)
+  }
+  case class ExternalVar(pack: PackageName, defName: Identifier)
+  extends NormalExpression {
+    val maxLambdaVar = None
+  }
+  case class Match(arg: NormalExpression,
+    branches: NonEmptyList[(Pattern[(PackageName, Constructor), Type], NormalExpression)])
+  extends NormalExpression {
+    val maxLambdaVar =
+      (arg.maxLambdaVar.toList ++ branches.toList.flatMap(_._2.maxLambdaVar))
+        .reduceLeftOption(Math.max)
+  }
+  case class LambdaVar(index: Int) extends NormalExpression {
+    val maxLambdaVar = Some(index)
+  }
+  /*
+   * It is reasonable to ask how you can define a lambda without an identifier
+   * for its argument in its expression. This is a benefit of de bruijn indexing.
+   * When all lambdas have exactly one argument you identify the var by how many
+   * lambdas out you have to travel.
+   *
+   * eg \x -> \y -> [y,x] would have normalization Lambda(Lambda(Apply(Apply(List, LambdaVar(1)), LambdaVar(0))))
+   *
+   * ref: https://en.wikipedia.org/wiki/De_Bruijn_index
+   */
+  case class Lambda(expr: NormalExpression) extends NormalExpression {
+    val maxLambdaVar = expr.maxLambdaVar.map(_ - 1)
+  }
+  case class Struct(enum: Int, args: List[NormalExpression]) extends NormalExpression {
+    val maxLambdaVar = args.flatMap(_.maxLambdaVar).reduceLeftOption(Math.max)
+  }
+  case class Literal(lit: Lit) extends NormalExpression {
+    val maxLambdaVar = None
+  }
+  case class Recursion(lambda: NormalExpression) extends NormalExpression {
+    val maxLambdaVar = lambda.maxLambdaVar
+  }
+}
+
+object Normalization {
+  case class NormalExpressionTag(ne: NormalExpression, children: Set[NormalExpression])
+  type NormalizedPM = PackageMap.Typed[(Declaration, Normalization.NormalExpressionTag)]
+  type NormalizedPac = Package.Typed[(Declaration, Normalization.NormalExpressionTag)]
+
+  private def normalOrderReduction(expr: NormalExpression): NormalExpression = {
+    import NormalExpression._
+    val nextExpr = expr match {
+      case App(Lambda(nextExpr), arg) => {
+        applyLambdaSubstituion(nextExpr, Some(arg), 0)
+      }
+      case App(fn, arg) => {
+        val nextFn = normalOrderReduction(fn)
+        App(nextFn, arg)
+      }
+      /*case Match(struct@Struct(enum, args), branches) => { TODO: If s matches a pattern we should perform lambdaSubstitution
+        Match(struct, branches)
+      }*/
+      case Recursion(Lambda(innerExpr)) if(innerExpr.maxLambdaVar.map(_ < 0).getOrElse(false)) => {
+        applyLambdaSubstituion(innerExpr, None, 0)
+      }
+      case Lambda(App(innerExpr, LambdaVar(0))) => {
+        innerExpr
```

Comment:
this is eta conversion, right? Can we add a comment to help me and other keep them straight.

Reply:
and a test to make sure it works